File: laboratories/ppo/exercises/lab07/php/src/Vehicles/Vehicle.php
<?php

namespace Ppo\Lab07\Vehicles;

abstract class Vehicle
{
    protected bool $isAnonymous = false;
    protected ?string $identifier;

    public function __construct(?string $identifier = null)
    {
        $this->identifier = $identifier;

        if (!$identifier) {
            $this->isAnonymous = true;
        }
    }

    final public function identify(): string
    {
        if ($this->isAnonymous) {
            return $this->getAnonymousIdentifier();
        }

        return $this->getIdentifier();
    }

    protected function getAnonymousIdentifier(): string
    {
        return "anonymous by {$this->getClass()}";
    }

    protected function getClass(): string
    {
        return strtolower(array_reverse(explode("\\", static::class))[0]);
    }

    abstract protected function getIdentifier(): string;
}

File: laboratories/ppo/exercises/lab07/python/src/vehicles/Vehicle.py
class Vehicle():
    is_anonymous = False
    identifier = None

    def __init__(self, identifier=None):
        self.identifier = identifier

        if identifier == None:
            self.is_anonymous = True

    def identify(self):
        if self.is_anonymous:
            return self._get_anonymous_identifier()

        return self._get_identifier()

    def _get_anonymous_identifier(self):
        return "anonymous by " + self.__class__.__name__.lower()

    def _get_identifier(self):
        raise Exception("method not implemented")
